### PR TITLE
[MM-13721] Fix the etag function GetEtagForProfilesNotInTeam

### PR DESCRIPTION
**Pull request by jfrerich**
#### Summary
The REST API endpoint GET /users has multiple Etags. The one for querying users not in a team is called GetEtagForProfilesNotInTeam. The query to resolve GetEtagForProfilesNotInTeam looks at the most recent `UpdateAt` for users not in the given team. 

However, if any but the most recent such user is added to the team, the etag won't change despite those users having a newer `UpdateAt`: because they won't be considered in the set anymore.

This PR provides a method to change the etag value if any user, including those not the most recently updated are removed from the query set.  

The solution appends the number of users, not in the teamId to the etag value.  If a user which does not have the latest `UpdateAt` value is added to the teamId, a GetEtagForProfilesNotInTeam, will return the latest UpdateAt value in the set and the appended value will be reduced by number of users added to the teamID.  Thus, the etag value will change.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-13721

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Added or updated unit tests (required for all new features)
- [ ] Added API documentation (required for all new APIs)
- [ ] All new/modified APIs include changes to the drivers
- [ ] Has enterprise changes (please link)
- [ ] Has UI changes
- [ ] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json)) updates
- [ ] Touches critical sections of the codebase (auth, upgrade, etc.)

### CODE REVIEW

**Comment by jfrerich** [comment]
## GetEtagForProfilesNotInTeam Testing Functionality (Psuedocode)

| line |  Command    |  Users not in teamid1  | Should Etag Change? | Current GetEtag Value (teamid1)   | Proposed GetEtag Value (teamId1) |
|--|----------------------|--------------|-----------------------|--------------------|-----------------------|
|1 |                      | <empty_set\> | `nil`                 |`nil`         | `nil`                 | 
|2 |u1.create             | u1           | `Y - new user`        |`u1.UpdateAt` | `concat(u1.UpdateAt,1)` |
|3 |u1.saveto(teamId1)    | <empty_set\> | `Y - No users`        |`nil`         | `nil`                 |
|4 |u2.create             | u2           | `Y - new user`        |`u2.UpdateAt` | `concat(u2.UpdateAt,1)` |
|5 |u2.saveto(teamId2)    | u2           | `N`                   |`u2.UpdateAt` | `concat(u2.UpdateAt,1)` |
|6 |u3.create             | u2, u3       | `Y - new user`        |`u3.UpdateAt` | `concat(u3.UpdateAt,2)` |
|7 |u2.saveto(teamId1)    | u3           | `Y - u2 joined team1` |`u3.UpdateAt` | `concat(u3.UpdateAt,1)` |
|8 |u2.UpdateAt(UpdateAt) | u3           | `N`                   |`u3.UpdateAt` | `concat(u3.UpdateAt,1)` |



## Failure Explanation
Consider the case that fails. 

**u2 created.  
u3 created.  
u2 added to TeamId1**

After u2 is added to TeamId1, GetEtagForProfilesNotInTeam(TeamId1) still returns u3.UpdateAt because u3 is the user with the latest UpdateAt value (from profiles NOT in TeamID1)  u2.UpdateAt Value is updated, but because it is added to TeamID1 it no longer shows up in the set of users NOT in TeamId1

## Proposed Solution
etag uses UpdateAt value from Latest Created User (Not in TeamId) + Number of Users (NOT in TeamId).

Again, consider the case that fails. 

**u2 created.  
u3 created.  
u2 added to TeamId1**

| line |  Command    |  Users not in teamid1  | Should Etag Change?  | Proposed GetEtag Value (teamId1) |
|------|-------------|------------------------|----------------------|----------------------------------|
| 1    |u2.create             | u2           | `Y - new user`        | `concat(u2.UpdateAt,1)` |
| 2    |u2.saveto(teamId2)    | u2           | `N`                   | `concat(u2.UpdateAt,1)` |
| 3    |u3.create             | u2, u3       | `Y - new user`        | `concat(u3.UpdateAt,2)` |
| 4    |u2.saveto(teamId1)    | u3           | `Y - u2 joined team1` | `concat(u3.UpdateAt,1)` |
| 5    |u2.UpdateAt(UpdateAt) | u3           | `N`                   | `concat(u3.UpdateAt,1)` |
| 6    |u4.create             | u3, u4       | `Y`                   | `concat(u4.UpdateAt,2)` |
| 7    |Remove(TeamID1, u2)   | u2, u3, u4   | `Y`                   | `concat(u2.UpdateAt,3)` |

Note on table line 4.  The etag value has changed because the query is aware that some previous user, NOT in TeamId1, is now in TeamID1.

**Cases covered by this method** 
1. Table line 6 - u4 is created, but not added to TeamID1. GetEtag will get updated.  `CONCAT(u4.UpdateAt, 2)`
2. Table line 7 - u2 is removed from TeamId1. GetEtag will get updated. `CONCAT(u2.UpdateAt, 3)`

**Comment by jfrerich** [comment]
**WIP**
- Need to optimize Query
- MySQL test passes, but Postgres fails

**Comment by lieut-data** [comment]
@jfrerich, excellent analysis. I suspect the historical reason for using `UpdateAt` in this case was to mirror other etag methodologies, where using `UpdateAt` might be overzealous (invalidating the etag too often) but would otherwise be correct.

I like the simplicity of your original proposal, but note that if one user is removed and another user is added to the team, the total count won't change and we're back to a variant of the original problem. Hashing the list would indeed guarantee correctness.

Is there a clever way to generate this hash in the database and so avoid the need to return all those users back to the application? Otherwise, my fear is that the cost of computing the etag that way would outweigh just calling `GetProfilesNotInTeam` in the first place.

**Comment by jfrerich** [comment]
@lieut-data My assumption was that adding or removing a user from a teamid would modify their UpdateAt value?  Is that not correct?

**Comment by lieut-data** [comment]
@jfrerich, you are correct: adding or removing a user from a team will touch their `UpdateAt`. Thinking through my comment more carefully, I do see that if we can assume `UpdateAt` always changes (increases) in tandem with a user entering or leaving the "not in team X" set, then your proposed etag implementation makes sense. 

There would be an edge case where the user is added at the same time as a user being removed. On our own community.mattermost.com, I note there are no users with duplicate `UpdateAt` fields. It seems like a reasonable tradeoff to adopt your approach vs. the complexity of hashing all the users not in a team.

File: store/sqlstore/user_store.go [hunk 1]
```diff
@@ -1347,23 +1347,20 @@ func (us SqlUserStore) GetProfilesNotInTeam(teamId string, offset int, limit int
...
-            ORDER BY u.UpdateAt DESC
-            LIMIT 1
-            `, map[string]interface{}{"TeamId": teamId})
 
+		var querystr string
+		querystr = `
+			SELECT 
+				CONCAT(MAX(UpdateAt), '.',  COUNT(Id)) as etag
+			FROM Users as u
+				LEFT JOIN TeamMembers tm ON tm.UserId=u.Id AND tm.TeamId=:TeamId AND tm.DeleteAt=0
```
Comment (lieut-data) [inline store/sqlstore/user_store.go]: From a code review perspective, can you adopt the SQL styling we're using elsewhere, basically putting each SQL statement on its own line, aligned with the previous lines, and with spaces around the `=` characters?

File: store/storetest/user_store.go [hunk 2]
```diff
@@ -2891,10 +2891,7 @@ func testUserStoreGetProfilesNotInTeam(t *testing.T, ss store.Store) {
 	store.Must(ss.Team().SaveMember(&model.TeamMember{TeamId: teamId, UserId: u2.Id}, -1))
 	u2.UpdateAt = store.Must(ss.User().UpdateUpdateAt(u2.Id)).(int64)
 
-	// GetEtagForProfilesNotInTeam only works if the most recent user is added to the team,
-	// otherwise the timestamp simply never changes: see https://mattermost.atlassian.net/browse/MM-13721.
 	t.Run("etag for profiles not in team 1 after update", func(t *testing.T) {
-		t.Skip()
```
Comment (lieut-data) [inline store/storetest/user_store.go]: 🎉

File: store/storetest/user_store.go [hunk 3]
```diff
@@ -2950,6 +2948,18 @@ func testUserStoreGetProfilesNotInTeam(t *testing.T, ss store.Store) {
 		etag4 := result.Data.(string)
 		require.Equal(t, etag3, etag4, "etag should not have changed")
 	})
+
+	// Add u3 to team 2
+	store.Must(ss.Team().SaveMember(&model.TeamMember{TeamId: teamId2, UserId: u3.Id}, -1))
+	u3.UpdateAt = store.Must(ss.User().UpdateUpdateAt(u3.Id)).(int64)
+
+	t.Run("etag for profiles not in team 1 after u3 added to team 2", func(t *testing.T) {
+		t.Skip()
```
Comment (lieut-data) [inline store/storetest/user_store.go]: Can you add a comment here explaining why we're skipping, and that the etag is still changing right now even though it wouldn't have to with a perfect hashing solution?

**Comment by jfrerich** [comment]
> Is there a clever way to generate this hash in the database and so avoid the need to return all those users back to the application? Otherwise, my fear is that the cost of computing the etag that way would outweigh just calling `GetProfilesNotInTeam` in the first place.

@lieut-data Sure.  One approach would be the following implementation for GetEtagForProfilesNotInteam():

1. Query the SORTED list of users, not in a team, from the database
2. generate sha1 hash value on string from step 1
3. return hash value as etag value.  

This method would:
- prevent falsely invalidating etags by removing everchanging UpdateAt dependency
- send only a small 40 digit hexadecimal sha1 value back to the application.
- reduce bandwidth for all false etag evaluations because GetProfilesNotInTeam will not be needed

The tradeoff is server-side compute time added from sha1 hash value generation versus bandwidth required for false invalid etags.

**Comment by lieut-data** [comment]
Agreed re: the algorithm above. If it could be all done database-server-side, it might be relatively quick: but would still require reading each row, whereas the current `UpdateAt` approach can leverage the index to pick out just the single row required. I suspect the overhead of the former to be non-trivial, and like your current approach best as a good compromise.

**Comment by jfrerich** [comment]
Thanks for your review and approval, @lieut-data !  I am researching the serverless-ops failed check right now.  It looks like I need to do a `git fetch && git merge origin/master` as opposed to a rebase?

**Comment by lieut-data** [comment]
Yep, that ought to do it!

**Pull request merged**